src/server/pps/server/pipeline_controller.go

```diff
@@ -123,7 +125,13 @@ func (m *ppsMaster) newPipelineOp(ctx context.Context, pipeline string) (*pipeli
 	}
 	// get latest PipelineInfo (events can pile up, so that the current state
 	// doesn't match the event being processed)
-	if err := m.a.pipelines.ReadOnly(ctx).Get(pipeline, op.pipelineInfo); err != nil {
+	if err := m.a.txnEnv.WithReadContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+		specCommit, err := m.a.findPipelineSpecCommit(txnCtx, pipeline, "")
+		if err != nil {
+			return err
+		}
+		return m.a.pipelines.ReadWrite(txnCtx.SqlTx).Get(specCommit, op.pipelineInfo)
```

Comment:
Isn't this a Read only transaction?

Reply:
We should make sure the we are passing the read only option to `dbutil.WithTx` so that we get the full benefits of a read only transaction, and also so that writing causes an error.

Reply:
handled these transactions in a more straightforward manner. Do you mean that for WithReadContext generally? We'd need to change how user-facing transactions work in that case since they rely on writes succeeding inside a "read only" transaction (just in the sense that it will not be committed). Probably not hard otherwise, though

Reply:
Okay, maybe add a comment saying that it's not actually a read only transaction.  But this doesn't need to change in this P.R.

Reply:
This bit is explicitly a ReadOnly access in the current version